core/consensus/babe/src/lib.rs

```diff
@@ -75,28 +155,918 @@ impl Decode for BabeSeal {
+	SO: SyncOracle + Send + Sync + Clone,
+	DigestItemFor<B>: CompatibleDigestItem + DigestItem<AuthorityId=Public>,
+	Error: ::std::error::Error + Send + From<::consensus_common::Error> + From<I::Error> + 'static,
+	OnExit: Future<Item=(), Error=()>,
+{
+	let worker = BabeWorker {
+		client: client.clone(),
+		block_import,
+		env,
+		local_key,
+		inherent_data_providers: inherent_data_providers.clone(),
+		sync_oracle: sync_oracle.clone(),
+		force_authoring,
+		threshold: config.threshold(),
+	};
+	slots::start_slot_worker::<_, _, _, _, _, BabeSlotCompatible, _>(
+		config.0,
+		client,
+		Arc::new(worker),
+		sync_oracle,
+		on_exit,
+		inherent_data_providers
+	)
+}
+
+struct BabeWorker<C, E, I, SO> {
+	client: Arc<C>,
+	block_import: Arc<I>,
+	env: Arc<E>,
+	local_key: Arc<sr25519::Pair>,
+	sync_oracle: SO,
+	inherent_data_providers: InherentDataProviders,
+	force_authoring: bool,
+	threshold: u64,
+}
+
+impl<B: Block, C, E, I, Error, SO> SlotWorker<B> for BabeWorker<C, E, I, SO> where
+	C: ProvideRuntimeApi + ProvideCache<B>,
+	C::Api: AuthoritiesApi<B>,
+	E: Environment<B, Error=Error>,
+	E::Proposer: Proposer<B, Error=Error>,
+	<<E::Proposer as Proposer<B>>::Create as IntoFuture>::Future: Send + 'static,
+	I: BlockImport<B> + Send + Sync + 'static,
+	SO: SyncOracle + Send + Clone,
+	DigestItemFor<B>: CompatibleDigestItem + DigestItem<AuthorityId=Public>,
+	Error: std::error::Error + Send + From<::consensus_common::Error> + From<I::Error> + 'static,
+{
+	type OnSlot = Box<Future<Item=(), Error=consensus_common::Error> + Send>;
+
+	fn on_start(
+		&self,
+		slot_duration: u64
+	) -> Result<(), consensus_common::Error> {
+		register_babe_inherent_data_provider(&self.inherent_data_providers, slot_duration)
+	}
+
+	fn on_slot(
+		&self,
+		chain_head: B::Header,
+		slot_info: SlotInfo,
+	) -> Self::OnSlot {
+		let pair = self.local_key.clone();
+		let _public_key = self.local_key.public();
+		let client = self.client.clone();
+		let block_import = self.block_import.clone();
+		let env = self.env.clone();
+
+		let (timestamp, slot_num, slot_duration) =
+			(slot_info.timestamp, slot_info.number, slot_info.duration);
+
+		let authorities = match authorities(client.as_ref(), &BlockId::Hash(chain_head.hash())) {
+			Ok(authorities) => authorities,
+			Err(e) => {
+				error!(
+					target: "babe",
+					"Unable to fetch authorities at block {:?}: {:?}",
+					chain_head.hash(),
+					e
+				);
+				telemetry!(CONSENSUS_WARN; "babe.unable_fetching_authorities";
+					"slot" => ?chain_head.hash(), "err" => ?e
+				);
+				return Box::new(future::ok(()));
+			}
+		};
+
+		if !self.force_authoring && self.sync_oracle.is_offline() && authorities.len() > 1 {
+			debug!(target: "babe", "Skipping proposal slot. Waiting for the network.");
+			telemetry!(CONSENSUS_DEBUG; "babe.skipping_proposal_slot";
+				"authorities_len" => authorities.len()
+			);
+			return Box::new(future::ok(()));
+		}
+
+		// FIXME replace the dummy empty slices with real data
```

Comment:
needs issue (probably should just be in the issue only and not clutter the code)

Reply:
This would be fixed before BABE works, so it would not stick around in the code, but whatever.

Reply:
yes, but it'll stick around until the next PR is in :)

Reply:
Issue filed